rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,208 @@
+# RFC 541 - 2022-12-09 - Reload API
+
+This RFC proposes adding an HTTP API for Vector's configuration reloading
+process. This will allow an enterprise control plane to more directly control a
+running instance and provide a path for synchronous runtime feedback that
+doesn't currently exist.
+
+## Context
+
+- [RFC: Observability Pipelines Worker (internal)](https://docs.google.com/document/d/10c6KuzbUcEUJxBfzX3a2CQaLHGPW2zKlz-_U8kS2DmA)
+- [Observability Pipelines Worker Implementation (internal)](https://docs.google.com/document/d/1hXv0fO1nsMkZeBIgZAsGD6QB9NACYaioxQ1FoQz94LI)
+
+## Cross cutting concerns
+
+- [Reloading RFC](https://docs.google.com/document/d/1rmLwFrx_yAeLHa_SeEyJx0Op6Z0EHpK6_LVa39fnlok)
+- [Add reload subcommand](https://github.com/vectordotdev/vector/issues/541)
+- [Data Samples for OP Configuration Builder (internal)](https://docs.google.com/document/d/1L8PUzDr1G9C3ORQPvMab6lcj-mh18P1nDq5ecnPf4mE)
+
+## Scope
+
+### In scope
+
+- Integrating an API that can initiate a reload and return success or failure
+- Refactoring Vector's main application loop to better support such an API
+- Allowing configuration error messages to be passed back via the API
+
+### Out of scope
+
+- Changing the actual reloading semantics of Vector in any way
+- Implementing APIs for any other type of control actions
+
+## Pain
+
+Our enterprise control plane component needs to initate configuration reloads
+in Vector and report back success or failure. Currently, the only way to do
+that would be to manipulate the config files on disk, ensure Vector is watching
+for changes, and parse the log output to determine whether or not the reload
+succeeded. This would be difficult to implement and likely unreliable.
+
+## Proposal
+
+### User Experience
+
+At startup, Vector will look for an environment variable
+(`VECTOR_CONTROL_SOCKET_PATH`). If that variable is present and contains a
+valid path, it will start an HTTP control server listening on a Unix socket at
+that path.
+
+The control server will support a single route at `POST /config`, which will
+accept a JSON-encoded version of a Vector configuration. Similar to config
+received from a provider implementation, input globs will be expanded but
+environment variables will not be interpolated. The HTTP response will
```

Comment:
That's a good question and one that I think applies more broadly to OPW. The simplest solution I think would be to have the API client do any interpolation on its side before building the config body into a request. But we'll likely want a more structured way of doing that than Vector's current raw substitution.

/cc @leebenson for any thoughts here

Reply:
If we do do the interpolation in the OPW, before passing to Vector for reload, we should do that for the initial start-up config too for consistency.

Reply:
Why can't environment variables be interpolated by Vector on reload? Does doing it in the API client/OPW amount to duplicating the Vector interpolation logic for now?

Reply:
I'd favour handling expansion in OPW:

1. We're likely to already be doing some preprocessing on configs (SDS -> remap potentially, removing/augmenting global config blocks, etc), so having OPW expand secrets and other config to form a 'final' config might already be relevant.
2. Secrets config expansion was rolled back because it broke the undocumented use case of using env vars for templating. It might be easier to bring if back if we prevent that use case in OPW.
3. We might need to expand env vars to drive Vector in some way (TBD).

Reply:
> Why can't environment variables be interpolated by Vector on reload?

The main technical limitation here is that uninterpolated variables (the way Vector currently implements interpolation, with raw string substitution) can make the config file invalid json/yaml/toml, or even just not parsable as a config directly due to type differences (e.g. int vs string). That's something we want to avoid sticking in a new API.

This is the same thing that caused issues with secrets expansion, so we're going to want a more structured solution for OP in general. Figuring out exactly what that should be is outside the scope of this RFC, so the goal here is to get something simple working that's functional enough for the first few use cases.